vsphere/virtual_machine_config_structure.go

```diff
@@ -0,0 +1,664 @@
+package vsphere
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"reflect"
+
+	"github.com/hashicorp/terraform/helper/logging"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/terraform-providers/terraform-provider-vsphere/vsphere/internal/helper/structure"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+var virtualMachineResourceAllocationTypeValues = []string{"cpu", "memory"}
+
+var virtualMachineVirtualExecUsageAllowedValues = []string{
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvAuto),
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvOn),
+	string(types.VirtualMachineFlagInfoVirtualExecUsageHvOff),
+}
+
+var virtualMachineVirtualMmuUsageAllowedValues = []string{
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageAutomatic),
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageOn),
+	string(types.VirtualMachineFlagInfoVirtualMmuUsageOff),
+}
+
+var virtualMachineSwapPlacementAllowedValues = []string{
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeInherit),
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeVmDirectory),
+	string(types.VirtualMachineConfigInfoSwapPlacementTypeHostLocal),
+}
+
+var virtualMachineFirmwareAllowedValues = []string{
+	string(types.GuestOsDescriptorFirmwareTypeBios),
+	string(types.GuestOsDescriptorFirmwareTypeEfi),
+}
+
+// getWithRestart fetches the resoruce data specified at key. If the value has
+// changed, a reboot is flagged in the virtual machine by setting
+// reboot_required to true.
+func getWithRestart(d *schema.ResourceData, key string) interface{} {
+	if d.HasChange(key) {
+		log.Printf("[DEBUG] %s: Resource argument %q requires a VM restart", resourceVSphereVirtualMachineIDString(d), key)
```

Comment:
`[DEBUG] %s: Resource` -> `[DEBUG] Virtual Machine %q: Resource`?

Reply:
The ID "stringer" function will take care of adding the VM and the ID.